File: scripts/log_setup.py
import atexit
import os
import re
import sys
import threading
import time
from datetime import datetime
from zoneinfo import ZoneInfo

from project_env import load_project_env


_LOG_SETUP_DONE = False
_ACTIVE_LOGGER = None
_CONFIGURED_SECRETS = ()
DEFAULT_LOG_RETENTION_DAYS = 30
DEFAULT_LOG_MAX_TOTAL_BYTES = 256 * 1024 * 1024
ACTIVE_LOG_GRACE_SECONDS = 5 * 60


def redact_log_text(value):
    text = str(value if value is not None else '')
    for secret in _CONFIGURED_SECRETS:
        text = text.replace(secret, '[REDACTED]')
    text = re.sub(
        r'\b([a-z][a-z0-9+.-]*://)([^\s/@]+)@',
        r'\1[REDACTED]@',
        text,
        flags=re.IGNORECASE,
    )
    credential_name = (
        r'(?:authorization|proxy-authorization|x-api-key|api[-_ ]?key|key|'
        r'access[-_ ]?token|refresh[-_ ]?token|token|secret|password|passwd|'
        r'cookie|set-cookie|paper_analyzer_api_key|kimi_api_key|'
        r'[a-z0-9-]+_(?:api_key|token|secret|password))'
    )
    text = re.sub(
        rf'((?:["\']?{credential_name}["\']?)\s*[:=]\s*)([^\r\n]+)',
        r'\1[REDACTED]',
        text,
        flags=re.IGNORECASE,
    )
    text = re.sub(r'\b(?:Bearer|Basic)\s+[A-Za-z0-9._~+/=-]+', '[REDACTED]', text, flags=re.IGNORECASE)
    text = re.sub(r'\bsk-[A-Za-z0-9._-]{3,}', '[REDACTED]', text, flags=re.IGNORECASE)
    return text


def format_log_timestamp(now=None):
    # The project uses Beijing time for all operational timestamps, regardless of host locale.
    if now is None:
        now = datetime.now(tz=ZoneInfo('Asia/Shanghai'))
    elif now.tzinfo is None:
        now = now.replace(tzinfo=ZoneInfo('Asia/Shanghai'))
    else:
        now = now.astimezone(ZoneInfo('Asia/Shanghai'))
    beijing = now
    return beijing.strftime('%Y-%m-%d %H:%M:%S.') + f'{beijing.microsecond // 1000:03d}+08:00'


class _Tee:
    def __init__(self, file_handle, original_stream, write_lock):
        self.file_handle = file_handle
        self.original_stream = original_stream
        self.write_lock = write_lock
        self.pending = ''

    @staticmethod
    def _format_complete_line(line):
        if line in ('\n', '\r\n'):
            return line
        ending = '\r\n' if line.endswith('\r\n') else '\n'
        content = line[:-len(ending)]
        return f'[{format_log_timestamp()}] {redact_log_text(content)}{ending}'

    def _write_output(self, text):
        if not text:
            return 0
        if self.file_handle is not None:
            self.file_handle.write(text)
            self.file_handle.flush()
        return self.original_stream.write(text)

    def write(self, data):
        text = str(data if data is not None else '')
        with self.write_lock:
            self.pending += text
            complete = self.pending.splitlines(keepends=True)
            if complete and not complete[-1].endswith(('\n', '\r')):
                self.pending = complete.pop()
            else:
                self.pending = ''
            rendered = ''.join(self._format_complete_line(line) for line in complete)
            self._write_output(rendered)
        return len(text)

    def drain(self):
        with self.write_lock:
            if not self.pending:
                return
            text = f'[{format_log_timestamp()}] {redact_log_text(self.pending)}'
            self.pending = ''
            self._write_output(text)

    def flush(self):
        with self.write_lock:
            if self.file_handle is not None:
                self.file_handle.flush()
            return self.original_stream.flush()

    def isatty(self):
        return self.original_stream.isatty()

    def fileno(self):
        return self.original_stream.fileno()


class _Logger:
    def __init__(self, file_handle, log_file, stdout, stderr):
        self.file_handle = file_handle
        self.log_file = log_file
        self.stdout = stdout
        self.stderr = stderr
        self.closed = False

    def close(self):
        if self.closed:
            return
        self.closed = True
        if isinstance(sys.stdout, _Tee) and sys.stdout.file_handle is self.file_handle:
            sys.stdout.drain()
        if isinstance(sys.stderr, _Tee) and sys.stderr.file_handle is self.file_handle:
            sys.stderr.drain()
        if isinstance(sys.stdout, _Tee) and sys.stdout.file_handle is self.file_handle:
            sys.stdout = self.stdout
        if isinstance(sys.stderr, _Tee) and sys.stderr.file_handle is self.file_handle:
            sys.stderr = self.stderr
        if self.file_handle is not None:
            try:
                self.file_handle.flush()
                os.fsync(self.file_handle.fileno())
            finally:
                self.file_handle.close()


def _timestamp():
    return datetime.now(tz=ZoneInfo('Asia/Shanghai')).strftime("%Y%m%d-%H%M%S")


def _create_unique_log_file(logs_dir, base_name):
    timestamp = _timestamp()
    for sequence in range(100):
        log_file = os.path.join(logs_dir, f"{base_name}-{timestamp}-{os.getpid()}-{sequence}.log")
        try:
            fd = os.open(log_file, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
            return os.fdopen(fd, 'w', encoding='utf-8', buffering=1), log_file
        except FileExistsError:
            continue
    raise RuntimeError(f"无法为 {base_name} 创建唯一日志文件")


def _positive_integer(value, fallback):
    try:
        parsed = int(str(value))
    except (TypeError, ValueError):
        return fallback
    return parsed if parsed > 0 else fallback


def _log_owner_process_is_alive(file_path):
    match = re.search(r'-(\d+)-\d+\.log$', os.path.basename(file_path))
    if not match:
        return False
    try:
        os.kill(int(match.group(1)), 0)
        return True
    except PermissionError:
        return True
    except (ProcessLookupError, ValueError, OverflowError):
        return False


def prune_log_files(logs_dir, *, retention_days=None, max_total_bytes=None, now=None):
    retention_days = _positive_integer(
        retention_days if retention_days is not None else os.environ.get('PD_LOG_RETENTION_DAYS'),
        DEFAULT_LOG_RETENTION_DAYS,
    )
    max_total_bytes = _positive_integer(
        max_total_bytes if max_total_bytes is not None else os.environ.get('PD_LOG_MAX_TOTAL_BYTES'),
        DEFAULT_LOG_MAX_TOTAL_BYTES,
    )
    now = time.time() if now is None else float(now)
    cutoff = now - retention_days * 24 * 60 * 60
    entries = []
    try:
        names = os.listdir(logs_dir)
    except OSError:
        return {'removed': 0, 'reclaimedBytes': 0}
    for name in names:
        if not name.endswith('.log'):
            continue
        file_path = os.path.join(logs_dir, name)
        try:
            stat = os.lstat(file_path)
        except OSError:
            continue
        if not os.path.isfile(file_path) or os.path.islink(file_path):
            continue
        entries.append({
            'path': file_path,
            'mtime': stat.st_mtime,
            'size': stat.st_size,
            'active_owner': _log_owner_process_is_alive(file_path),
        })

    removed = 0
    reclaimed = 0
    retained = []
    for entry in entries:
        if not entry['active_owner'] and entry['mtime'] < cutoff:
            try:
                os.unlink(entry['path'])
                removed += 1
                reclaimed += entry['size']
                continue
            except FileNotFoundError:
                continue
            except OSError:
                pass
        retained.append(entry)
    retained.sort(key=lambda item: (-item['mtime'], item['path']))
    total_bytes = sum(item['size'] for item in retained)
    for entry in reversed(retained):
        if total_bytes <= max_total_bytes:
            break
        # Do not unlink a log that another still-running process may own.
        if entry['active_owner'] or entry['mtime'] >= now - ACTIVE_LOG_GRACE_SECONDS:
            continue
        try:
            os.unlink(entry['path'])
        except FileNotFoundError:
            total_bytes -= entry['size']
            continue
        except OSError:
            continue
        total_bytes -= entry['size']
        removed += 1
        reclaimed += entry['size']
    return {'removed': removed, 'reclaimedBytes': reclaimed}


def setup_script_logging(script_path=None):
    global _LOG_SETUP_DONE, _ACTIVE_LOGGER, _CONFIGURED_SECRETS
    if _LOG_SETUP_DONE:
        return _ACTIVE_LOGGER
    _LOG_SETUP_DONE = True
    load_project_env()
    _CONFIGURED_SECRETS = tuple(
        str(value)
        for key, value in os.environ.items()
        if re.search(r'(?:API_KEY|SECRET|TOKEN|PASSWORD|PASSWD|COOKIES?)$', key, re.IGNORECASE)
        and len(str(value)) >= 6
    )
    disable_file_logs = (
        os.environ.get("PAPER_DIGEST_DISABLE_FILE_LOGS") == "1"
        or os.environ.get("PD_DISABLE_FILE_LOGS") == "1"
    )

    if not script_path:
        script_path = sys.argv[0] if sys.argv else "script.py"

    scripts_dir = os.path.dirname(os.path.abspath(__file__))
    project_root = os.path.dirname(scripts_dir)
    logs_dir = os.path.join(project_root, "logs")
    base_name = os.path.splitext(os.path.basename(script_path))[0] or "script"
    fh = None
    log_file = None
    if not disable_file_logs:
        os.makedirs(logs_dir, exist_ok=True)
        prune_log_files(logs_dir)
        fh, log_file = _create_unique_log_file(logs_dir, base_name)

    original_stdout = sys.stdout
    original_stderr = sys.stderr
    write_lock = threading.RLock()
    sys.stdout = _Tee(fh, original_stdout, write_lock)
    sys.stderr = _Tee(fh, original_stderr, write_lock)
    _ACTIVE_LOGGER = _Logger(fh, log_file, original_stdout, original_stderr)
    atexit.register(_ACTIVE_LOGGER.close)
    if log_file:
        print(f"[log] 输出文件: {log_file}")
    return _ACTIVE_LOGGER

File: scripts/log-setup.js
const fs = require('fs');
const path = require('path');
const { loadProjectEnv } = require('./env-loader.js');

let activeLogger = null;
let fileSequence = 0;
let configuredSecrets = [];
const DEFAULT_LOG_RETENTION_DAYS = 30;
const DEFAULT_LOG_MAX_TOTAL_BYTES = 256 * 1024 * 1024;
const ACTIVE_LOG_GRACE_MS = 5 * 60 * 1000;

function ensureDir(dirPath) {
    if (!fs.existsSync(dirPath)) {
        fs.mkdirSync(dirPath, { recursive: true });
    }
}

function getBeijingTimeParts(date = new Date(), fractionalSecondDigits) {
    const options = {
        timeZone: 'Asia/Shanghai',
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
        hour: '2-digit',
        minute: '2-digit',
        second: '2-digit',
        hourCycle: 'h23'
    };
    if (fractionalSecondDigits) options.fractionalSecondDigits = fractionalSecondDigits;
    return Object.fromEntries(
        new Intl.DateTimeFormat('en-CA', options)
            .formatToParts(date)
            .filter(part => part.type !== 'literal')
            .map(part => [part.type, part.value])
    );
}

function formatTs(date = new Date()) {
    const values = getBeijingTimeParts(date);
    return `${values.year}${values.month}${values.day}-${values.hour}${values.minute}${values.second}`;
}

function formatLogTimestamp(date = new Date()) {
    const values = getBeijingTimeParts(date, 3);
    return `${values.year}-${values.month}-${values.day} ${values.hour}:${values.minute}:${values.second}.${values.fractionalSecond}+08:00`;
}

function timestampLogLines(text, state) {
    let output = '';
    for (const char of text) {
        if (state.atLineStart && char !== '\n' && char !== '\r') {
            output += `[${formatLogTimestamp()}] `;
            state.atLineStart = false;
        }
        output += char;
        if (char === '\n') state.atLineStart = true;
    }
    return output;
}

function redactLogText(value) {
    let text = String(value ?? '');

    for (const secret of configuredSecrets) {
        if (secret && text.includes(secret)) text = text.split(secret).join('[REDACTED]');
    }

    // URL credentials must be removed before generic credential fields are handled.
    text = text.replace(
        /\b([a-z][a-z0-9+.-]*:\/\/)([^\s/@]+)@/gi,
        '$1[REDACTED]@'
    );

    // Header, environment-variable and JSON-style credential fields.
    text = text.replace(
        /((?:["']?(?:authorization|proxy-authorization|x-api-key|api[-_ ]?key|key|access[-_ ]?token|refresh[-_ ]?token|token|secret|password|passwd|cookie|set-cookie|paper_analyzer_api_key|kimi_api_key|[a-z0-9-]+_(?:api_key|token|secret|password))["']?)\s*[:=]\s*)([^\r\n]+)/gi,
        '$1[REDACTED]'
    );

    // Also protect standalone authorization values and commonly printed key fragments.
    text = text.replace(/\b(?:Bearer|Basic)\s+[A-Za-z0-9._~+/=-]+/gi, '[REDACTED]');
    text = text.replace(/\bsk-[A-Za-z0-9._-]{3,}/gi, '[REDACTED]');

    return text;
}

function setStdoutBlocking() {
    if (process.stdout._handle && process.stdout._handle.setBlocking) {
        process.stdout._handle.setBlocking(true);
    }
}

function isTestProcess() {
    if (process.env.NODE_TEST_CONTEXT) return true;
    return process.argv.some(arg => /(?:^|[/\\])tests[/\\].+\.test\.js$/.test(arg));
}

function createUniqueLogFile(logsDir, base) {
    for (let attempt = 0; attempt < 100; attempt += 1) {
        const sequence = fileSequence++;
        const suffix = `${formatTs()}-${process.pid}-${sequence}`;
        const logFile = path.join(logsDir, `${base}-${suffix}.log`);
        try {
            const fd = fs.openSync(logFile, 'ax', 0o600);
            return { fd, logFile };
        } catch (err) {
            if (err.code !== 'EEXIST') throw err;
        }
    }
    throw new Error(`无法为 ${base} 创建唯一日志文件`);
}

function readPositiveInteger(value, fallback) {
    const parsed = Number.parseInt(String(value ?? ''), 10);
    return Number.isInteger(parsed) && parsed > 0 ? parsed : fallback;
}

function logOwnerProcessIsAlive(filePath) {
    const match = path.basename(filePath).match(/-(\d+)-\d+\.log$/);
    if (!match) return false;
    const pid = Number.parseInt(match[1], 10);
    if (!Number.isInteger(pid) || pid <= 0) return false;
    try {
        process.kill(pid, 0);
        return true;
    } catch (error) {
        return error?.code === 'EPERM';
    }
}

function pruneLogFiles(logsDir, options = {}) {
    const retentionDays = readPositiveInteger(
        options.retentionDays ?? process.env.PD_LOG_RETENTION_DAYS,
        DEFAULT_LOG_RETENTION_DAYS
    );
    const maxTotalBytes = readPositiveInteger(
        options.maxTotalBytes ?? process.env.PD_LOG_MAX_TOTAL_BYTES,
        DEFAULT_LOG_MAX_TOTAL_BYTES
    );
    const nowMs = Number.isFinite(options.nowMs) ? options.nowMs : Date.now();
    const cutoffMs = nowMs - retentionDays * 24 * 60 * 60 * 1000;
    let entries = [];
    try {
        entries = fs.readdirSync(logsDir, { withFileTypes: true }).flatMap(entry => {
            if (!entry.isFile() || !entry.name.endsWith('.log')) return [];
            const filePath = path.join(logsDir, entry.name);
            try {
                const stat = fs.lstatSync(filePath);
                if (!stat.isFile() || stat.isSymbolicLink()) return [];
                return [{
                    filePath, mtimeMs: stat.mtimeMs, size: stat.size,
                    activeOwner: logOwnerProcessIsAlive(filePath)
                }];
            } catch (_) {
                return [];
            }
        });
    } catch (_) {
        return { removed: 0, reclaimedBytes: 0 };
    }

    const remove = entry => {
        try {
            fs.unlinkSync(entry.filePath);
            return true;
        } catch (error) {
            return error.code === 'ENOENT';
        }
    };
    let removed = 0;
    let reclaimedBytes = 0;
    const retained = [];
    for (const entry of entries) {
        if (!entry.activeOwner && entry.mtimeMs < cutoffMs && remove(entry)) {
            removed += 1;
            reclaimedBytes += entry.size;
        } else {
            retained.push(entry);
        }
    }
    retained.sort((a, b) => b.mtimeMs - a.mtimeMs || a.filePath.localeCompare(b.filePath));
    let totalBytes = retained.reduce((sum, entry) => sum + entry.size, 0);
    for (const entry of retained.slice().reverse()) {
        if (totalBytes <= maxTotalBytes) break;
        // A concurrently running script may still own a newly touched file.
        // Let capacity temporarily exceed the target instead of unlinking it.
        if (entry.activeOwner || entry.mtimeMs >= nowMs - ACTIVE_LOG_GRACE_MS) continue;
        if (!remove(entry)) continue;
        totalBytes -= entry.size;
        removed += 1;
        reclaimedBytes += entry.size;
    }
    return { removed, reclaimedBytes };
}

function normalizeWriteArgs(chunk, encoding, callback) {
    let resolvedEncoding = encoding;
    let resolvedCallback = callback;
    if (typeof encoding === 'function') {
        resolvedCallback = encoding;
        resolvedEncoding = undefined;
    }
    const source = Buffer.isBuffer(chunk)
        ? chunk.toString(typeof resolvedEncoding === 'string' ? resolvedEncoding : 'utf8')
        : String(chunk);
    return {
        text: redactLogText(source),
        encoding: resolvedEncoding,
        callback: resolvedCallback
    };
}

function setupScriptLogging(scriptPath, options = {}) {
    if (activeLogger) return activeLogger;

    // A non-default env file is accepted only through this explicit test/programmatic API.
    loadProjectEnv(options.envFile);
    configuredSecrets = Object.entries(process.env)
        .filter(([key, value]) => /(?:API_KEY|SECRET|TOKEN|PASSWORD|PASSWD|COOKIES?)$/i.test(key) && String(value).length >= 6)
        .map(([, value]) => String(value));
    setStdoutBlocking();

    if (isTestProcess() && !options.allowInTestProcess) return null;
    const disableFileLogs = process.env.PAPER_DIGEST_DISABLE_FILE_LOGS === '1'
        || process.env.PD_DISABLE_FILE_LOGS === '1';

    const entryPath = scriptPath || process.argv[1] || __filename;
    const projectRoot = path.resolve(__dirname, '..');
    const logsDir = options.logsDir || path.join(projectRoot, 'logs');
    const base = path.basename(entryPath, path.extname(entryPath)) || 'script';
    let fd = null;
    let logFile = null;
    if (!disableFileLogs) {
        ensureDir(logsDir);
        pruneLogFiles(logsDir);
        ({ fd, logFile } = createUniqueLogFile(logsDir, base));
    }
    const stdoutWrite = process.stdout.write.bind(process.stdout);
    const stderrWrite = process.stderr.write.bind(process.stderr);
    let fileOpen = fd !== null;
    let fileFailed = false;
    let closing = false;
    let closePromise = null;

    function reportFileError(err) {
        if (fileFailed) return;
        fileFailed = true;
        const message = timestampLogLines(
            redactLogText(`[log] 文件日志写入失败: ${err.message}\n`),
            { atLineStart: true }
        );
        try {
            stderrWrite(message);
        } catch (_) {
            // Logging failure must not terminate the business script.
        }
    }

    function wrapWrite(originalWrite) {
        const state = { atLineStart: true };
        return (chunk, encoding, callback) => {
            const args = normalizeWriteArgs(chunk, encoding, callback);
            args.text = timestampLogLines(args.text, state);
            if (fileOpen && !fileFailed && !closing) {
                try {
                    fs.writeSync(fd, args.text, null, typeof args.encoding === 'string' ? args.encoding : 'utf8');
                } catch (err) {
                    reportFileError(err);
                }
            }
            return originalWrite(args.text, args.encoding, args.callback);
        };
    }

    process.stdout.write = wrapWrite(stdoutWrite);
    process.stderr.write = wrapWrite(stderrWrite);

    function restoreWrites() {
        process.stdout.write = stdoutWrite;
        process.stderr.write = stderrWrite;
    }

    function flushAndClose() {
        if (!fileOpen) return;
        fileOpen = false;
        if (!fileFailed) {
            try {
                fs.fsyncSync(fd);
            } catch (err) {
                reportFileError(err);
            }
        }
        if (fd !== null) {
            try {
                fs.closeSync(fd);
            } catch (err) {
                reportFileError(err);
            }
        }
    }

    function close() {
        if (closePromise) return closePromise;
        closing = true;
        restoreWrites();
        flushAndClose();
        closePromise = Promise.resolve();
        return closePromise;
    }

    activeLogger = { logFile, close };
    global.__PAPER_DIGEST_LOG_SETUP_DONE__ = true;

    process.once('beforeExit', close);
    process.once('exit', () => {
        restoreWrites();
        flushAndClose();
    });

    if (logFile) console.log(`[log] 输出文件: ${logFile}`);
    return activeLogger;
}

function closeScriptLogging() {
    return activeLogger ? activeLogger.close() : Promise.resolve();
}

module.exports = {
    setupScriptLogging,
    closeScriptLogging,
    redactLogText,
    setStdoutBlocking,
    formatTs,
    formatLogTimestamp,
    timestampLogLines,
    pruneLogFiles,
    DEFAULT_LOG_RETENTION_DAYS,
    DEFAULT_LOG_MAX_TOTAL_BYTES,
    ACTIVE_LOG_GRACE_MS
};
